src/cljam/algo/pileup.clj

```diff
@@ -9,29 +9,45 @@
   (:import [cljam.io.protocols SAMAlignment]
            [cljam.io.pileup PileupBase LocusPile]))
 
+(def ^:const base-chunk-size 5000)
+
 (defn- cover-locus? [^long pos ^SAMAlignment aln]
   (<= (.pos aln) pos))
 
-(defn pileup-seq
-  "Returns a lazy sequence that each element contains [position (alignments
-  piled-up at the locus)]."
-  [^long start ^long end alns]
+(defn- seq-step
+  [^long start ^long end ^long step-size alns]
   (letfn [(step [^long pos prev-buf rest-alns]
             (lazy-seq
-             (when (< (dec pos) end)
-               (let [[i rests] (split-with (partial cover-locus? pos) rest-alns)
-                     p (some-> rests ^SAMAlignment first .pos)
-                     [b :as buf] (into (filterv #(<= pos (.end ^SAMAlignment %))
-                                                prev-buf) i)
-                     next-pos (if (and p (not b)) p (inc pos))]
-                 (if b
-                   (cons [pos buf] (step next-pos buf rests))
-                   (when (and p (<= p end))
-                     (step next-pos buf rests)))))))]
+              (when (< (dec pos) end)
+                (let [[i rests] (split-with (partial cover-locus?
+                                                     (+ pos step-size))
+                                            rest-alns)
+                      p (some-> rests ^SAMAlignment first .pos)
```

Comment:
Wow, I didn't know there are cases where type hints may be ignored only when direct linking is enabled. Looks like a Clojure compiler bug. 🤔

Reply:
Fixed

Reply:
Thanks!

I figured out that `^SAMAlignment first` tags the var `first` as a function which returns a `SAMAlignment` instance while `^SAMAlignment (first)` tags `(first rests)` as a form that is evaluated to be a `SAMAlignment` instance.

When direct linking is enabled, the tag on the var is ignored because `((VarExpr)fexpr).tag` is not used while parsing.

- https://github.com/clojure/clojure/blob/75f8bc98b4e8d46dbaa28e54adfa24946e314bbe/src/jvm/clojure/lang/Compiler.java#L3688-L3689
- https://github.com/clojure/clojure/blob/75f8bc98b4e8d46dbaa28e54adfa24946e314bbe/src/jvm/clojure/lang/Compiler.java#L3844

I don't know if this is a bug or not since the behavior of locally overwriting return type hints of functions is not documented...